### PR TITLE
Actvate MMU on Aarch64

### PR DESCRIPTION
**Pull request by jamesmintram**
This goal of this PR is to activate the MMU in the Aarch64 prekernel. Peripheral address ranges are retrieved from the MMIO singleton. Only the first 1GB of main memory is mapped in the perkernel.

The prekernel is getting bigger, so I did some refactoring to make the prekernel init easier to follow.

This PR has been tested in Qemu and on a physical Rpi3B+ and Rpi4B

### CODE REVIEW

File: Kernel/Prekernel/Arch/aarch64/init.cpp [hunk 1]
```diff
@@ -6,22 +6,17 @@
...
-extern "C" [[noreturn]] void init();
-extern "C" [[noreturn]] void os_start();
-
-static void set_up_el1_mode();
-static void set_up_el2_mode();
-static void set_up_el3_mode();
-static void print_current_exception_level(const char* msg);
-[[noreturn]] static void jump_to_os_start_from_el2();
-[[noreturn]] static void jump_to_os_start_from_el3();
+extern "C" void wait_cycles(int n);
```
[the remark this answers is not in the document]
Comment (jamesmintram) [reply Kernel/Prekernel/Arch/aarch64/init.cpp:23]: This function already exists with that signature, so I would consider that change out of scope of this PR.

File: Kernel/Prekernel/Arch/aarch64/PrekernelMMU.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    {
+        if (m_start >= m_end) {
+            Prekernel::panic("Invalid memory range passed to PageBumpAllocator");
+        }
+        if ((FlatPtr)m_start % PAGE_TABLE_SIZE != 0 || (FlatPtr)m_end % PAGE_TABLE_SIZE != 0) {
+            Prekernel::panic("Memory range passed into PageBumpAllocator not aligned to PAGE_TABLE_SIZE");
+        }
+    }
+
+    uint64_t* TakePage()
```
[the remark this answers is not in the document]
Comment (jamesmintram) [reply Kernel/Prekernel/Arch/aarch64/PrekernelMMU.cpp]: I have renamed `TakePage` -> `take_page()` 

I have used `FlatPtr` to signify a pointer stored as an u64. I assume this is the correct usage?

I looked at PhysicalPage and PhysicalAddress and they don't seem appropriate here. One is ref counted, the other does a lot more than we need. 

> m_current by (PAGE_TABLE_SIZE / sizeof(FlatPtr)) 

That is the amount we need to bump out u64* ptr by to "step over" 1 page table's worth of memory.

File: Kernel/Arch/aarch64/Registers.h [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+
+        return spsr;
+    }
+};
+static_assert(sizeof(SPSR_EL2) == 8);
+
+// https://developer.arm.com/documentation/ddi0595/2021-06/AArch64-Registers/SPSR-EL3--Saved-Program-Status-Register--EL3-
+// Saved Program Status Register
+struct SPSR_EL3 {
+    enum Mode : uint16_t {
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Kernel/Arch/aarch64/Registers.h:408]: missed use of stdint.h in this file

**Review by ADKaster — APPROVED** [review]
Very nice changes now. One missed comment of mine from earlier but otherwise looks good to go

**Review by ADKaster — APPROVED** [review]
Looks good. One final thing: typo in the linker script changes. idntity --> identity. If I don't comment on it I guarantee Linus or Andreas would 🙃. Otherwise looks like a great step towards getting the processor in a properly booted state.

**Comment by jamesmintram** [comment]
> Looks good. One final thing: typo in the linker script changes. idntity --> identity. If I don't comment on it I guarantee Linus or Andreas would upside_down_face. Otherwise looks like a great step towards getting the processor in a properly booted state.

Resolved

**Comment by bgianfo** [comment]
Hey @jamesmintram I was about to merge this, but then I tried to build it locally.. but the aarch64 compilation fails:

```
/PrekernelCommon.cpp.o -c ../../Kernel/Prekernel/Arch/aarch64/PrekernelCommon.cpp
../../Kernel/Prekernel/Arch/aarch64/PrekernelCommon.cpp:9:10: fatal error: Kernel/Arch/aarch64/Aarch64Asm.h: No such file or directory
    9 | #include <Kernel/Arch/aarch64/Aarch64Asm.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Can you take a look?

**Comment by bgianfo** [comment]
Can confirm it's now compiling and "boots" to logo in qemu.

LGTM, thanks! :rocket:

**Pull request merged**